src/client/sandbox/node/window.js

```diff
@@ -16,44 +16,57 @@ import constructorIsCalledWithoutNewKeyword from '../../utils/constructor-is-cal
 const nativeFunctionToString = nativeMethods.Function.toString();
 
 export default class WindowSandbox extends SandboxBase {
-    constructor (nodeSandbox, messageSandbox) {
+    constructor (nodeSandbox, messageSandbox, listenersSandbox) {
         super();
 
-        this.nodeSandbox    = nodeSandbox;
-        this.messageSandbox = messageSandbox;
+        this.nodeSandbox      = nodeSandbox;
+        this.messageSandbox   = messageSandbox;
+        this.listenersSandbox = listenersSandbox;
 
         this.UNCAUGHT_JS_ERROR_EVENT   = 'hammerhead|event|uncaught-js-error';
+        this.UNHANDLED_REJECTION_EVENT = 'hammerhead|event|unhandled-rejection';
         this.FORCE_PROXY_SRC_FOR_IMAGE = 'hammerhead|image|force-proxy-src-flag';
     }
 
-    _raiseUncaughtJsErrorEvent (msg, window, pageUrl) {
+    _raiseUncaughtJsErrorEvent (type, msg, window, pageUrl) {
         if (!isCrossDomainWindows(window, window.top)) {
             const sendToTopWindow = window !== window.top;
 
             if (!pageUrl)
                 pageUrl = destLocation.get();
 
             if (sendToTopWindow) {
-                this.emit(this.UNCAUGHT_JS_ERROR_EVENT, {
-                    msg,
-                    pageUrl,
+                this.emit(type, { msg, pageUrl, inIframe: true });
+                this.messageSandbox.sendServiceMsg({ msg, pageUrl, cmd: type }, window.top);
+            }
+            else
+                this.emit(type, { msg, pageUrl });
+        }
+    }
 
-                    inIframe: true
-                });
+    _reattachUnhandledRejectionHandler (window) {
+        nativeMethods.windowRemoveEventListener.call(window, 'unhandledrejection', this);
+        nativeMethods.windowAddEventListener.call(window, 'unhandledrejection', this);
+    }
 
-                this.messageSandbox.sendServiceMsg({
-                    msg,
-                    pageUrl,
+    handleEvent (event) {
+        if (event.type === 'unhandledrejection' && !event.defaultPrevented) {
+            const reason = event.reason;
```

Comment:
Move to the separate function `WindowSandbox.formatUnhandledRejectionReason (eventReason)` and covers it by client tests